lua/telescope/actions/init.lua

```diff
@@ -1055,6 +1057,260 @@ actions.which_key = function(prompt_bufnr, opts)
+  end, smart)
+  return selected
+end
+
+local is_dir = function(value)
+  return string.sub(value, -1, -1) == os_sep
+end
+
+--- Creates a new file in the current directory of the |builtin.file_browser|.
+--- Notes:
+--- - You can create folders by ending the name in the path separator of your OS, e.g. "/" on Unix systems
+--- - You can implicitly create new folders by passing $/CWD/new_folder/filename.lua
+---@param prompt_bufnr number: The prompt bufnr
+actions.create_file = function(prompt_bufnr)
+  local current_picker = action_state.get_current_picker(prompt_bufnr)
+  local finder = current_picker.finder
+  local file = action_state.get_current_line()
+  if file == "" then
+    print(
+      "To create a new file or directory(add "
+        .. os_sep
+        .. " at the end of file) "
+        .. "write the desired new into the prompt and press <C-e>. "
+        .. "It works for not existing nested input as well."
+        .. "Example: this"
+        .. os_sep
+        .. "is"
+        .. os_sep
+        .. "a"
+        .. os_sep
+        .. "new_file.lua"
+    )
+    return
+  end
+  local fpath = Path:new(finder.path .. os_sep .. file)
+  if fpath:exists() then
+    print "File or folder already exists."
+    return
+  end
+  if not is_dir(fpath.filename) then
+    Path:new(fpath):touch { parents = true }
+    current_picker:refresh(false, { reset_prompt = true })
+  else
+    -- TODO(conni2461): I think when doing realpath we don't have to worry about :sub(1, -2) anymore
+    Path:new(fpath.filename:sub(1, -2)):mkdir { parents = true }
+    -- local new_cwd = vim.loop.fs_realpath(fpath)
+  end
+  current_picker:refresh(false, { reset_prompt = true })
+end
+
+-- creds to nvim-tree.lua
+local function rename_loaded_buffers(old_name, new_name)
+  for _, buf in pairs(a.nvim_list_bufs()) do
+    if a.nvim_buf_is_loaded(buf) then
+      if a.nvim_buf_get_name(buf) == old_name then
+        a.nvim_buf_set_name(buf, new_name)
+        -- to avoid the 'overwrite existing file' error message on write
+        vim.api.nvim_buf_call(buf, function()
+          vim.cmd "silent! w!"
+        end)
+      end
+    end
+  end
+end
+
+--- Rename files or folders for |builtin.file_browser|.<br>
+---@param prompt_bufnr number: The prompt bufnr
+actions.rename_file = function(prompt_bufnr)
+  local current_picker = action_state.get_current_picker(prompt_bufnr)
+  local entry = action_state.get_selected_entry()
+  local index = current_picker.manager:find_entry(entry)
+  local row = current_picker:get_row(index)
+  local old_name = Path:new(entry[1])
+
+  -- "../" more common so test first
+  if old_name.filename == "../" or old_name.filename == "./" then
+    print "Please select a file!"
+    return
+  end
+  entry[1] = vim.fn.input("Insert a new name: ", old_name:make_relative())
+
+  -- avoid error on refresh
+  if old_name:is_dir() and entry[1]:sub(-1, -1) ~= os_sep then
+    entry[1] = entry[1] .. os_sep
```

Comment:
I didn't like the result of this. Before rename I see filename under selection, after rename I see fullpath? it should just update the filename and that's it.

Also, I believe a print message from old_filename to new_filename would be important here.

Reply:
Ahh, yeah. Thanks for spotting this :) I haven't yet renamed the right file :laughing:  It's probably better done with picker refresh again.